### PR TITLE
Add auspice magic link endpoint

### PR DESCRIPTION
**Pull request by lvreynoso**
### Summary:
- **What:** Adds endpoints to stream Phylo Tree json data directly to Auspice via a magic link.
- **Ticket:** [sc184540](https://app.shortcut.com/genepi/story/184540)
- **Env:** [magiclink-frontend.dev.czgenepi.org](https://magiclink-frontend.dev.czgenepi.org)

### Checklist:
- [x] I merged latest `<base branch>`
- [x] I manually verified the change
- [x] I added labels to my PR
- [x] I tested in multiple browsers
- [x] I added relevant unit tests
- [x] I have notified others of changes they need to make locally (migrations, jobs, package updates, etc)

### CODE REVIEW

File: src/backend/aspen/api/views/auspice.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+    json_data = _rename_nodes_on_tree([json_data["tree"]], identifier_map, "GISAID_ID")
+
+    return json_data
+
+
+# TODO: Convert to POST request
+@router.post("/generate")
+async def generate_auspice_string(
+    request: Request,
```
Comment (jgadling) [inline src/backend/aspen/api/views/auspice.py:119]: this request can be a GenerateAuspiceMagicLinkRequest and fastapi/pydantic will validate the input automatically for you

File: src/backend/aspen/api/views/auspice.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        "expiry": expiry_time.isoformat(),
+    }
+
+    # encode before hashing
+    bytes_payload = json.dumps(payload).encode("utf8")
+    message = urlsafe_b64encode(bytes_payload)
+    # our mac key is stored as a b64encoded string
+    mac_key = urlsafe_b64decode(settings.AUSPICE_MAC_KEY)
+    digest_maker = hmac.new(mac_key, message, hashlib.sha3_512)
+    mac_tag = digest_maker.hexdigest()
```
Comment (jgadling) [inline src/backend/aspen/api/views/auspice.py:148]: We don't have to do this right now, but eventually we probably want to add these sign/authenticate code blocks to a shared library.

File: src/backend/aspen/api/views/auspice.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    return GenerateAuspiceMagicLinkResponse.parse_obj(
```
Comment (jgadling) [inline src/backend/aspen/api/views/auspice.py]: We don't have to use parse_obj here. We can run:
```return GenerateAuspiceMagicLinkResponse(url=f'url_stuff_here')```

File: src/backend/aspen/api/views/auspice.py (hunk 1 above)
Comment (lvreynoso) [reply src/backend/aspen/api/views/auspice.py:119]: True, but we need access to `request.url.netloc` to generate the url we are returning. Is there a way to make a pydantic schema preserve that information?

File: src/backend/aspen/api/utils/authz.py [hunk 4]
```diff
@@ -62,3 +62,30 @@ def authz_samples_cansee(
         and_(or_(~Sample.private, Sample.submitting_group_id == user.group_id))
     )
     return query
+
+
+# TODO, this is incredibly similar to sample authz filters. Generalize these!
```
Comment (jgadling) [inline src/backend/aspen/api/utils/authz.py:67]: ❤️

File: src/backend/aspen/api/views/auspice.py (hunk 1 above)
Comment (jgadling) [reply src/backend/aspen/api/views/auspice.py:119]: Fair point. It might be possible to do it other ways but this seems fine for now.

**Review by jgadling — APPROVED** [review]
Approved!

Please add unique MAC keys to staging and prod when you get a moment!

**Comment by lvreynoso** [comment]
keys added!

File: src/backend/aspen/api/views/auspice.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    payload_message, payload_mac_tag = magic_link.split(".")
+    encoded_payload_message = payload_message.encode("utf8")
+
+    mac_key = urlsafe_b64decode(settings.AUSPICE_MAC_KEY)
+    digest_maker = hmac.new(mac_key, encoded_payload_message, hashlib.sha3_512)
+    correct_mac_tag = digest_maker.hexdigest()
+
+    authenticated = hmac.compare_digest(payload_mac_tag, correct_mac_tag)
+    if not authenticated:
+        raise ex.BadRequestException(
```
Comment (vincent-czi) [inline src/backend/aspen/api/views/auspice.py:172]: It might be good for this to also kick off a report to Sentry.

**Review by vincent-czi — COMMENTED** [review]
Looks great! I really like how the magic link works!

**Pull request merged**